src/detours.cpp

```diff
@@ -334,14 +362,72 @@ struct _DETOUR_TRAMPOLINE
     PBYTE           pbRemain;       // first instruction after moved code. [free list]
     PBYTE           pbDetour;       // first instruction of detour function.
     BYTE            rbCodeIn[8];    // jmp [pbDetour]
+#ifdef DETOURS_SUPPORT_CONTEXT_CALLBACK
+    BYTE            rbContext[42];  // context callback code (calls context function w/ void* param)
+    PBYTE           pbContext;      // first instruction of context callback code (rbSetContext) if used
+    PF_DETOUR_INVOKE_CONTEXT_CALLBACK pfContextCallback;    // context callback used as indirect jmp in rbContext
+#endif // DETOURS_SUPPORT_CONTEXT_CALLBACK
 };
 
+#ifdef DETOURS_SUPPORT_CONTEXT_CALLBACK
+C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 160);
+#else
 C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 96);
+#endif // DETOURS_SUPPORT_CONTEXT_CALLBACK
 
 enum {
     SIZE_OF_JMP = 5
 };
 
+#ifdef DETOURS_SUPPORT_CONTEXT_CALLBACK
+
+inline PBYTE detour_gen_context_callback(PBYTE pbCode, PF_DETOUR_INVOKE_CONTEXT_CALLBACK* ppfContextCallback, PVOID pContextParam)
+{
+    PBYTE pbJmpSrc = pbCode + 26;
+
+    // Backup volatile registers that are used to pass __fastcall params.
+    // If the set context function performed floating point operations, 
+    // registers XMM0, XMM1, XMM2, and XMM3 would also have to be backed 
+    // up and restored. In the attempt to keep this code generation as simple 
+    // as possible, floating point operations in the set context function isn't 
+    // currently supported.
+    *pbCode++ = 0x51;   // push rcx (backup rcx)
```

Comment:
Is this the start of a function?
Are you aware of the ABI detail that functions must start with instructions that are at least one byte?
Well, it is for hotpatch, maybe no matter.